### PR TITLE
start drawing program and corresponding test file

### PR DESCRIPTION
**Pull request by lemireg**
wrote the first few methods for drawing program and the skeleton of the other methods. made notes of which tests we need to run for drawing program

### CODE REVIEW

**Comment by lemireg** [comment]
Question Sean wanted to ask Varik: "_putting drawing program iterator into the drawing program. (esp if we already have a list of shapes in there)_"

File: draw/shape_factory.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+# e.g.: my_shape = ShapeFactory.create_shape("Circle", 2.0), where 2.0 is the radius of the circle
+# e.g.: my_shape = ShapeFactory.create_shape("Rectangle", 2.0, 4.0), where 2.0 is length and 4.0 is width
+# place the @staticmethod decorator/annotation on top of your create_shape definition to allow you to call it as shown in above examples
+# NOTE: If you would like, you can instead have individual create methods for circle, square, triangle, and rectangle.
+# This will get rid of the need for a big nested if else inside create_shape. Thus you could have create_circle, create_square, etc.
+#
+
+class ShapeFactory(Circle, Square, Rectangle, Triangle):
+
+    def __init(self):
```
Comment (warlicks) [inline draw/shape_factory.py:13]: I think we can get a way with not creating an init method. ShapeFactory is really just going to call one of the shape classes and return a shape object. 

I am happy to jump in here and work on `ShapeFactory` and write the corresponding tests.

File: draw/DrawingProgram.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
+# Your team will create a program that has a class that holds a list/collection of shapes. This class will
+# be named DrawingProgram. Here are the attributes (data) and behaviors (methods) of this class:
+from shape_factory import ShapeFactory
+
+
+class DrawingProgram(ShapeFactory):
```
Comment (warlicks) [inline draw/DrawingProgram.py:6]: Do we actually  need to inherit from `ShapeFactory`. I think we are just using its capabilities.

**Comment by warlicks** [comment]
@lemireg I did get a response from Varik. He does want us to build a separate iterator class and to write the sort algorithm ourselves.

**Comment by warlicks** [comment]
> Question Sean wanted to ask Varik: "_putting drawing program iterator into the drawing program. (esp if we already have a list of shapes in there)_"

@lemireg I did get a response from Varik. He does want us to build a separate iterator class and to write the sort algorithm ourselves.

File: draw/shape_factory.py (hunk 1 above)
Comment (warlicks) [reply draw/shape_factory.py:13]: See #5

File: draw/DrawingProgram.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        else:
+            raise Exception(f'{shape} not in {self.shapes}')
+
+        print(f"{shape_rm_count} number of occurrences removed from shape list.")
+        return shape_rm_count
+
+    def print_shapes(self, shape):
+        """prints all shapes that match the type of the shape passed in.
+        throw exception if shape is not in ShapeFactory shapes."""
+        if shape not in ShapeFactory.shape_dict():
```
Comment (warlicks) [inline draw/DrawingProgram.py:36]: I broke this when I reworked the Shapefactory

File: draw/DrawingProgram.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+    def get_shape(self, index):
+        """returns the shape at the specified index"""
+
+    def set_shape(self, index, shape):
+        """replaces the shape at the specified index any other
+        behaviors you feel are necessary for the class"""
+
+
+class DrawingProgramIterator:
```
Comment (warlicks) [inline draw/DrawingProgram.py:65]: Would it be helpful if I took a swing at this?

File: draw/DrawingProgram.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        else:
+            Exception(f"Shapes: {shapes} is not None or a list.")
+
+    def add_shape(self, shape):
+        self.shapes.append(shape)
+
+    def remove_shape(self, shape):
+        """remove shape that exists in the DrawingProgram.Shapes list, else throw an exception"""
+        shape_rm_count = self.shapes.count(shape)
+        if shape in self.shapes:
```
Comment (warlicks) [inline draw/DrawingProgram.py:25]: Need to access the shape_name or check the shape class.

File: draw/DrawingProgram.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+class DrawingProgramIterator:
+    """DrawingProgramIterator provides the ability to iterate
+    across the collection of shapes in DrawingProgram using a for loop."""
+    # See notes/slides on Iterator and
+    # book chapter "The Iterator Pattern" and the section "The Iterator Protocol"
+    # Sean's question: (why not build it into the class?)
+
+
+class DrawingProgramMain(DrawingProgram):
```
Comment (warlicks) [inline draw/DrawingProgram.py:73]: I'm going to purpose breaking this out of this file. I think this file will get unmanageable if we add this in there too.

File: draw/DrawingProgram.py (hunk 4 above)
Comment (warlicks) [reply draw/DrawingProgram.py:65]: I will work on this tonight (11/21) and just add it to #5

File: draw/DrawingProgram.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+class DrawingProgram(ShapeFactory):
+    # For the intents and purposes of this assignment, all shapes should be created via ShapeFactory.
+    #     Thus any adding of shapes to your DrawingProgram class object must be done through ShapeFactory.
+    #     This may seem cumbersome, but the point is to leave object creation to a single class.
+    #         It encapsulates how to create the objects you need.
+    #  ^^^ I think this means we pass in ShapeFactory into DrawingProgram instead of the shapes themselves?
+
+    def __init__(self, shapes = None):
+        if shapes is None or isinstance(shapes, list):
```
Comment (warlicks) [inline draw/DrawingProgram.py:14]: We need it initiate `self.shapes` as an empty list for other things to work down stream. For example if we try use `.append()` when self.shapes is `None` we will get an error. 
I'm going to fix this as I write the `DrawingProgramIterator`. I'll make it an empyt list by default.

**Pull request merged**